packages/ide-extension/src/telemetry/telemetry.ts

```diff
@@ -0,0 +1,148 @@
+        const client = new TelemetryClient(key);
+        client.channel.setUseDiskRetryCaching(true);
+        client.addTelemetryProcessor((envelope: Contracts.Envelope) => {
+            envelope.tags['ai.location.ip'] = '0.0.0.0';
+            envelope.tags['ai.cloud.roleInstance'] = 'masked';
+            return true;
+        });
+        client.context.tags[client.context.keys.userId] = env.machineId;
+        client.context.tags[client.context.keys.sessionId] = env.sessionId;
+        client.context.tags[client.context.keys.cloudRole] = env.appName;
+        const enabled = updateTelemetryStatus();
+        disposables.push(workspace.onDidChangeConfiguration(() => updateTelemetryStatus()));
+        reporter = {
+            client,
+            enabled,
+            dispose: () => {
+                disposables.forEach((d) => d.dispose());
+                reporter = undefined;
+            }
+        };
+    }
+    return reporter;
+}
+
+/**
+ * Update the telemetry setting by reading configuration.
+ *
+ * @returns - status of telemetry setting, true: enabled; false: disabled
+ */
+function updateTelemetryStatus(): boolean {
+    const enabled = !!workspace.getConfiguration('sap.ux.knowledgeHub').get('telemetryEnabled');
+    if (reporter) {
+        reporter.enabled = enabled;
+    }
+    return enabled;
+}
+
+/**
+ * Set common properties which will be added to every telemetry event.
+ * If called without properties, all common properties are removed.
+ *
+ * @param properties - name/value pair of properties (optional)
+ * @param properties.ide - development environment VSCODE or SBAS
+ * @param properties.devSpace - SBAS devspace
+ */
+export function setCommonProperties(properties?: { ide: 'VSCODE' | 'SBAS'; devSpace: string }) {
+    if (reporter) {
+        reporter.commonProperties = properties
+            ? {
+                  'cmn.appstudio': properties.ide === 'SBAS' ? 'true' : 'false',
+                  'cmn.devspace': properties.devSpace,
+                  apiHost: '',
+                  apiVersion: '',
+                  'common.os': platform(),
+                  'common.nodeArch': arch(),
+                  'common.platformversion': (release() || '').replace(/^(\d+)(\.\d+)?(\.\d+)?(.*)/, '$1$2$3'),
+                  'common.extname': packageJson.name,
+                  'common.extversion': packageJson.version
+              }
+            : undefined;
+    }
+}
+
+/**
+ * Track an even using telemetry.
+ *
+ * @param event - telemetry event
+ */
+export async function trackEvent(event: TelemetryEvent): Promise<void> {
+    if (!reporter?.enabled) {
+        return;
+    }
+    try {
+        const name = `${packageJson.name}/${event.name}`;
+        const properties = propertyValuesToString({ ...event.properties, ...(reporter.commonProperties || {}) });
+        reporter.client.trackEvent({ name, properties });
+        logString(`Telemetry event '${event.name}': ${JSON.stringify(event.properties)}`);
```

Comment:
Is this required? Does this need to be behind a `debug` log level?

Reply:
@kranthie-sap I could not understand this one. Are you referring to the logString function or all 4 lines?

Reply:
Fixed.